feat: add TLS and RFC 1929 auth to KumoProxy

vietcgi opened the pull request:
## Summary

- Add TLS encryption support for KumoProxy SOCKS5 server connections
- Implement RFC 1929 username/password authentication for client access control
- Add client-side TLS support in kumod for connecting to TLS-enabled proxies
- Support mTLS (mutual TLS) for client certificate verification

Fixes #451

## Changes

**Proxy Server (crates/proxy-server):**
- New `--tls-certificate`, `--tls-private-key` options for server TLS
- New `--tls-client-ca` option for mTLS client verification
- New `--auth-file` option for TOML-based user credentials
- Generic async stream handling (works with both plain TCP and TLS)
- Constant-time password comparison to prevent timing attacks

**KumoMTA Client (crates/kumod):**
- New `socks5_proxy_username` and `socks5_proxy_password` fields
- New `socks5_proxy_tls`, `socks5_proxy_tls_ca`, `socks5_proxy_tls_insecure` fields
- TLS connector with optional CA verification and insecure mode

**Documentation:**
- Added docs for all new `make_egress_source` options

## Test plan

- [x] Unit tests for auth module (9 tests pass)
- [x] Manual testing on Linux (Ubuntu 24.04 via multipass):
  - Plain TCP + NOAUTH
  - Plain TCP + Auth (valid/invalid credentials)
  - TLS + NOAUTH
  - TLS + Auth
  - Auth rejection with wrong password
  - TLS mismatch (plain client -> TLS server)
- [x] cargo check, cargo clippy pass
- [x] Backward compatible (no auth/TLS required by default)

wez commented on crates/kumod/src/egress_source.rs:
```diff
@@ -973,17 +1030,152 @@ impl<'a> ProxyProto<'a> {
+}
+
+/// Build a TLS connector for SOCKS5 proxy connections.
+async fn build_socks5_tls_connector(
```
I'd like to avoid copying and pasting the other existing and honestly rather *complex* tls code here, as it will lead to subtle divergence issues in the future.

What I propose is extracting `crates/rfc5321/src/tls.rs` into a separate crate, eg: `crates/kumo-tls-helper`.

I think that `crates/rfc5321/src/traits.rs` should probably also be moved there, as the two are coupled.

Then the logic here for the socks client should be re-phrased to work in terms of the `TlsOptions` struct exposed by that new `kumo-tls-helper` crate.

wez commented on crates/proxy-server/src/tls_helpers.rs:
```diff
@@ -0,0 +1,104 @@
+//! TLS configuration helpers for the proxy server.
```
by switching to using `kumo-server-common` as proposed above, we can then simply reference its `make_server_config` function directly rather than copying the code into this file.

wez commented on crates/kumo-tls-helper/src/lib.rs:
```diff
@@ -61,7 +88,7 @@ impl std::hash::Hash for RustlsCacheKey {
 lruttl::declare_cache! {
 /// Caches TLS connector information for the RFC5321 SMTP client
-static RUSTLS_CACHE: LruCacheWithTtl<RustlsCacheKey, Arc<ClientConfig>>::new("rfc5321_rustls_config", 32);
+static RUSTLS_CACHE: LruCacheWithTtl<RustlsCacheKey, Arc<ClientConfig>>::new("kumo_tls_helper_rustls_config", 32);
```
Renaming this cache is effectively an API change, and we should mention it in the changelog under breaking changes in case someone has a policy that is tuning the cache capacity.

The list of caches is automatically extracted and displayed in https://docs.kumomta.com/reference/kumo/set_lruttl_cache_capacity/ when the docs are built.

I think renaming this makes sense, as its scope is potentially increased as part of moving it.

I'd suggest naming it `rustls_client_config`.

```suggestion
static RUSTLS_CACHE: LruCacheWithTtl<RustlsCacheKey, Arc<ClientConfig>>::new("rustls_client_config", 32);
```

wez commented on crates/proxy-server/src/main.rs:
```diff
@@ -1,81 +1,290 @@
+    // Legacy CLI options for backwards compatibility
+    // These are mutually exclusive with --policy
+    /// [Legacy] Address(es) to listen on (e.g., "0.0.0.0:5000").
+    /// Mutually exclusive with --policy.
```
You can do something like:

```rust
#[arg(long, conflicts_with="proxy-config")]
```

to have the clap argument parser enforce the mutually exclusive/conflicting arguments.

Be sure to check that this works as you expect: I don't know if it wants `proxy-config` or `proxy_config`, and you won't know which until you try to run it--it won't catch an issue at compile time.

The same can be applied to each of the conflicting arguments.

wez commented on crates/proxy-server/src/main.rs:
```diff
@@ -1,81 +1,290 @@
+}
+
+/// Run in legacy mode using CLI options (backwards compatibility)
+async fn run_legacy(opts: Opt) -> anyhow::Result<()> {
```
I would love to avoid this code duplication!

My suggestion is that when the user doesn't provide a path to a config script, that we create one based on the command line arguments:

* Use `tempfile::NamedTempFile::with_prefix("kumo-proxy-config")` to create a temporary file
* We write out to it based on what we see in `opts`:

```lua
local kumo = require 'kumo'

kumo.on('proxy_init', function()
```

Then loop over the `opts.listen` and emit appropriate listener entries:

```
  kumo.start_proxy_listener {
    listen = LISTENER,
    -- no_splice and timeout here too   
  }
```

and then close it out:

```
end)
```

With that populated, we can then pass in that path via `StartConfig::policy`.

wez commented on crates/proxy-server/src/mod_proxy.rs:
```diff
@@ -0,0 +1,221 @@
+#[derive(Clone, Debug, serde::Serialize)]
+pub struct ConnMeta {
+    pub peer_address: String,
+    pub local_address: String,
```
you can have serde take care of serializing these to strings

```suggestion
    pub peer_address: SockAddr,
    pub local_address: SockAddr,
```

wez commented on crates/integration-tests/src/test/proxy_basic.rs:
```diff
@@ -0,0 +1,335 @@
+
+    let response = MailGenParams {
+        ..Default::default()
+    }
```
```suggestion
    let response = MailGenParams::default()
```

wez reviewed (commented):
Thank you for your work on this, I'm excited to see this get merged!

It took me a little while to get back to you on this next round of review because I was building out the new ACL system with this in mind and wanted to have that in a better state before suggesting here how to integrate those changes here.

vietcgi commented:
Hey @wez,
  
I've addressed all your feedback from today's review:

Switched to AuthInfo return type with the new ACL system integration
Using wait_for_source_summary instead of sleep in tests + proper delivery assertions
Consolidated the lua configs into just proxy_with_auth.lua
Reverted the maildir-sink.lua changes
All the small suggestions (MailGenParams::default, error handling cleanup, etc)

I squashed it down to 2 clean commits. Ready for another look when you get a chance!

vietcgi commented on crates/integration-tests/proxy_auth.lua:
sure we can

wez commented on crates/integration-tests/src/test/proxy_basic.rs:
```diff
@@ -0,0 +1,365 @@
+    // Wait for delivery
+    daemon
+        .kumod
+        .wait_for_maildir_count(1, Duration::from_secs(10))
```
```suggestion
        .wait_for_maildir_count(1, Duration::from_secs(50))
```

wez commented on crates/integration-tests/src/test/proxy_basic.rs:
```diff
@@ -0,0 +1,365 @@
+        .kumod
+        .wait_for_source_summary(
+            |summary| summary.get(&TransientFailure).copied().unwrap_or(0) > 0,
+            Duration::from_secs(10),
```
```suggestion
            Duration::from_secs(50),
```

wez commented on crates/integration-tests/src/test/proxy_basic.rs:
```diff
@@ -0,0 +1,365 @@
+
+    // Verify transient failures occurred (require_auth rejected the client)
+    let delivery_summary = daemon.kumod.dump_logs().await?;
+    anyhow::ensure!(
```
please also assert that the delivery count is 0 here too

wez commented on crates/proxy-server/src/main.rs:
```diff
@@ -1,81 +1,157 @@
+            .proxy_config
+            .unwrap_or_else(|| PathBuf::from("/opt/kumomta/etc/proxy/init.lua"));
+        (path, None)
+    };
```
It's slightly more idiomatic and cleaner to flip this around this way; we can remove the impossible `unwrap_or_else` case and its fallback to a default path that we'll never use:

```suggestion
    let legacy_file;
    
    let policy_path = match opts.proxy_config {
        Some(path) => path,
        None => {
           // Legacy mode: generate a temporary Lua config from CLI options
           legacy_file = opts.generate_legacy_config()?;
           temp_file.path().to_path_buf()
        }
    };
```

That `let legacy_file;` line looks weird on its own up there at first, but it is the idiomatic way to hold ownership of the temporary file object and avoid adding boilerplate/noise to thread an optional extra thing around through the match.

wez commented on crates/proxy-server/src/mod_proxy.rs:
```diff
@@ -0,0 +1,244 @@
+/// containing peer_address and local_address.
+/// Returns an AuthInfo on success (with the authenticated identity populated),
+/// or None if authentication failed.
+pub async fn authenticate_user(
```
I'm normally in favor of breaking things out into separate functions like this, but I think that in this case it leads to more code than is necessary; I think that we'll have clearer and more concise code if this logic was inlined into the `perform_auth` function so that it could then directly consult the `AuthKindResult` and either communicate the authentication failure directly, or directly update its `auth_kind` state.

wez reviewed (commented):
Thank you!

I noticed when checking it out and building locally that `splice(2)` wasn't hooked up, so I added some code that should work to enable that again.  It compiles, but I haven't explicitly verified that it it is operating correctly.  I force-pushed my updates to your branch.

Could you take a look to verify that splice is operating correctly? This doesn't strictly require an integration test (although I would appreciate it if you feel like making one!).  To verify I suppose that you could inspect the log output to see that splice is being used, and then also check to make sure that we're not leaking any additional file descriptors while the service is running and handling connections.  I would suggest looking at `lsof -np $(pgrep proxy-server)` and comparing splice to no-splice mode.

With that verified, the only thing remaining is to write up a doc page for the new `kumo.start_proxy_listener` function, then add an entry to the changelog for it!

Thank you again for your work and attention here!

wez commented:
Thank you! I've merged this. I made a couple of fixups to the docs, and I also renamed the `no_splice` option in lua to `use_splice` to avoid negative logic.  The command line argument remains `--no-splice`.

I think the only thing left to do now before we can consider #451 complete is to enable using TLS on the client side; then we can document the proxy user/pass and TLS options.

Would you mind opening a PR for that?

Thank you again!
